### PR TITLE
fix(#1966): go2Gno loses type info

### LINKED ISSUE #1966: go2Gno loses type info
see this:
```go
package main

import (
	"fmt"
	"time"
)

func main() {
	var a int64 = 2
	//fmt.Println(a * time.Second)
	fmt.Println(time.Second * a)
}
```

this would not compile in Go, but works in gno. the reason for it is while converting from go to gno, time.Duration is treated as a int64 rather than the time.Duration, the declared type is lost.

### PR DESCRIPTION
**Pull request by omarsy**
fixes #1966 

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by thehowl** [comment]
cc @ltzmaxwell

File: gnovm/pkg/gnolang/preprocess.go [hunk 1]
```diff
@@ -882,17 +882,26 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 					// Left not const, Right not const ------------------
 					if n.Op == EQL || n.Op == NEQ {
 						// If == or !=, no conversions.
-					} else if lnt, ok := lt.(*NativeType); ok {
+					} else if ok, lnt := func() (bool, *NativeType) {
```
Comment (deelawn) [inline gnovm/pkg/gnolang/preprocess.go:885]: Can you simplify this condition to something like this?
```go
} else if lnt, ok := lt.(*NativeType); ok && isNative(rt) {
```

Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go]: right it is better. Thx ^^

commit: https://github.com/gnolang/gno/pull/2016/commits/a8e4e7d3065205616836cc20913fef0278be6ba9

**Comment by ltzmaxwell** [comment]
thank you for your work, start looking.

File: gnovm/pkg/gnolang/preprocess.go [hunk 2]
```diff
@@ -882,17 +882,22 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
...
-						// If left and right are native type,
+						// If left and right are native type, and same type
 						// convert left and right to gno, then
 						// convert result back to native.
 						//
 						// get concrete native base type.
+						if lt.TypeID() != rt.TypeID() {
+							panic(fmt.Sprintf(
+								"incompatible types in binary expression: %v %v %v",
+								n.Left, n.Op, n.Right))
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:899]: maybe print the type info is more straightforward, what do you think?

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:899]: And, can you add a file test to better display this error message?

File: gnovm/pkg/gnolang/preprocess.go [hunk 3]
```diff
@@ -882,17 +882,22 @@ func Preprocess(store Store, ctx BlockNode, n Node) Node {
 					// Left not const, Right not const ------------------
 					if n.Op == EQL || n.Op == NEQ {
 						// If == or !=, no conversions.
-					} else if lnt, ok := lt.(*NativeType); ok {
+					} else if lnt, ok := lt.(*NativeType); ok && isNative(rt) {
```
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/preprocess.go:885]: this seems wrong. when Lhs is native and Rhs not native, the logic inside the `else if` block is skipped, and the Lhs will not be converted to gno type, so the left type and right type will never match.. right?

Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go:885]: I believe that when Lhs is native and Rhs is not native, they are not of the same type, which should result in an error being raised.

Referring to the comment at https://github.com/gnolang/gno/blob/master/gnovm/pkg/gnolang/preprocess.go#L891-L892, where we write that we should only enter the 'else if' condition when lt is native and rt is Native. Do you have any scenarios in mind where lt is native and rt is not native, but we should not raise an error?

File: gnovm/pkg/gnolang/preprocess.go (hunk 2 above)
Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go:899]: Yes right, I will do it

File: gnovm/pkg/gnolang/preprocess.go (hunk 3 above)
Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:885]: make sense to me, ty.

File: gnovm/pkg/gnolang/preprocess.go (hunk 2 above)
Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go:899]: Added one more test here: https://github.com/gnolang/gno/pull/2016/commits/e6e8d333c593deb210f812f4f5770a0fd65b7e92

> maybe print the type info is more straightforward, what do you think?

What do you mean here ?

Comment (ltzmaxwell) [reply gnovm/pkg/gnolang/preprocess.go:899]:
```go
panic(fmt.Sprintf("incompatible types in binary expression: %v %v %v",lt.TypdID(), n.Op, rt.TypeID()))
															
```
what do you think?

Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess.go]: I don't have an opinion on this. I don't know why we did it before (but the typeID printout is clearer).

Done here: https://github.com/gnolang/gno/pull/2016/commits/959b43e5bc7ff886586e499262b76644a8e0b1df

**Comment by ltzmaxwell** [comment]
cc @deelawn for a second look.

**Review by thehowl — APPROVED** [review]
Please fix up the unit tests; after that we're good to go.

**Review by petar-dambovaliev — APPROVED** [review]

File: gnovm/pkg/gnolang/preprocess_test.go
[the remark this answers is not in the document]
Comment (omarsy) [reply gnovm/pkg/gnolang/preprocess_test.go:1]: Thx for your review. ^^
Done here: https://github.com/gnolang/gno/pull/2016/commits/0148ad8c1e7e45ab20090857cf0f715af2171dce

**Review by thehowl — APPROVED** [review]

**Pull request merged**